### ISSUE TITLE
Linux terminal hangs after quit the GUI on Net5.0

### ISSUE THREAD

**Issue by BDisp** [migueldeicaza/gui.cs#931]
Run `UICatalog `in `WSL `and quit. Try to type and it's hang and you have to close `WSL`. Do the same with the `Example `and you can type normally. May be it's related with `Shutdown `already related before.

**Comment by tig** [comment]
I can't repro this either in Windows Terminal with Ubuntu or ConEmu.

What terminal are you using?

**Comment by tig** [comment]
That said, almost nothing else works in WSL for me. Tab key works to move between panes, but arrows don't nav. Hmm....

**Comment by tig** [comment]
Also not working with CMD and WSL2.

**Comment by tig** [comment]
`Example` has same issue for me.

**Comment by BDisp** [comment]
I use the default WSL2 terminal. I'm not sure what which open. How can I check that? I'm going test in VMWare to see if it's fail too. It's all messed up. With the Ubuntu 20.04 I seen the Example is fail too.

**Comment by tig** [comment]
Ahhh... 

If I start it from bash it works fine (and exits fine). but if I run it from pwsh most keystrokes don't work (but it exits fine).

**Comment by tig** [comment]
Weird thing is, `ocgv` works fine everywhere!

**Comment by BDisp** [comment]
The hangs after quit only happens with the Net5.0. The mouse issues only happens with Ubuntu which has the libncurses 6.2 version.

**Comment by BDisp** [comment]
Writing `reset `or `stty sane`, even if I can't see it, the terminal returns to normal. The problem is to code this in `ncurses`.

**Comment by BDisp** [comment]
I already discovered what is causing this. Are the `StartReportingMouseMoves `and `StopReportingMouseMoves `methods.
```

		public override void StartReportingMouseMoves ()
		{
			Console.Out.Write ("\x1b[?1003h");
			Console.Out.Flush ();
		}

		public override void StopReportingMouseMoves ()
		{
			Console.Out.Write ("\x1b[?1003l");
			Console.Out.Flush ();
		}

```
Commenting them the behavior issue stops but the problem it's the mouse movement also stops too.

**Comment by BDisp** [comment]
I won't waste any more time on this problem because it only happens with `Net5.0`. With `netcoreapp3.1` this problem does not happen, so until the final version of `Net5.0` this problem can be corrected.

**Comment by markusr** [comment]
The problem still exists. I just tested with ```Net5.0``` and ```gui.cs 1.0.0-pre.6``` .

**Comment by BDisp** [comment]
Yes indeed. But if you test with the NetDriver with the Application.UseSystemConsole = true the hangs don't happens, neither on Windows, WSL or Linux.

**Comment by markusr** [comment]
I confirm that it does not hang any more. However, after I close the application and I use a mouse button in the terminal I get strange text like:
```�,`�(`�(a�(a�(a�(a�(a�(`�-`�-`�.`�.a�.a�.a�.`�.`�.`�.`�.a�.a�.a�.a�.a�.a``` 
on the terminal.

My environment is: ``` Kubuntu 20.04, Konsole Version 19.12.3```

**Comment by BDisp** [comment]
Ah I see. I am using `CSI `in the `Unix `for resizing and perhaps it is not clearing on exit.

**Comment by markusr** [comment]
Is this another bug or related? Shall I open another issue?

**Comment by BDisp** [comment]
No, it is still under development in this PR #1021.

**Comment by rgl** [comment]
I just did a `dotnet build -c Release && ./UICatalog/bin/Release/net5.0/UICatalog` of 62b5800c84bb5b22935f2e63dc75db1b47bbadae on Ubuntu 20.04 gnome-terminal and after quitting the application, the terminal is fubar.

**Comment by BDisp** [comment]
Comment the `Application.HeightAsBuffer = true;` or set `Application.HeightAsBuffer = false;` and see the difference, please. Thanks.

**Comment by rgl** [comment]
Tried both, but it did not fix the problem.

**Comment by BDisp** [comment]
Can you post a print screen, please? Thanks.

**Comment by rgl** [comment]
Oh, I'm now noticing that I didn't explain the problem before... sorry! After the application quits, when I press the left (or right) cursor key, the cursor caret moves a whole word at a time instead of a character at a time. It seems the `ctrl` key is stuck. Hummm, I'm also noticing this after running the `dotnet` command... maybe this is a problem with the dotnet runtime? anyhow, whatever it is, its messing the terminal.

**Comment by BDisp** [comment]
Ah. Ok. The terminal is not restoring properly. It is a issue not solved yet.

**Comment by BDisp** [comment]
To bypass this without close the terminal, I do:
Ctrl+C
reset

With `Application.UseSystemConsole`, running with the `-usc` argument that behaviour don't happens.

**Comment by markusr** [comment]
Yes. My terminal is broken after
```dotnet build -c Release && ./UICatalog/bin/Release/net5.0/UICatalog```
then ```CTRL+Q```
Tested with master.

**Comment by migueldeicaza** [comment]
tagging myself, so I can review later
@migueldeicaza

**Comment by tig** [comment]
I can no longer repro this in WSL (Debian) from `main`. 

I can no longer repro this in WSL (Ubuntu 20.04) from `main`.

Anyone else?

**Comment by BDisp** [comment]
On WSL (Ubuntu 20.04.2 LTS) and on a Linux (VirtualBox) still happens with `Curses`. I have to type `reset` to restore.
With the `NetDriver` there is no hang.
Tested with the latest `main`.

**Comment by tig** [comment]
> On WSL (Ubuntu 20.04.2 LTS) and on a Linux (VirtualBox) still happens with `Curses`. I have to type `reset` to restore.
> With the `NetDriver` there is no hang.
> Tested with the latest `main`.

Can you repro it with WSL as well?

**Comment by BDisp** [comment]
Yes @tig

**Comment by tig** [comment]
I wonder what's different on my machine.

**Comment by BDisp** [comment]
Maybe your Linux version (Debian) but I doubt. It's not affecting only to me, because others are already say the same. It's only happens using he `CursesDriver`.
I already tried to resolve but without success. I only only run through SSH with remote access to shared folders. I don't build inside Linux.

**Comment by Artemigos** [comment]
I can also reproduce this on linux with both alacritty and xterm. After quitting the app, the terminal keeps adding characters with just the mouse moving over it. Even in the app the mouse behaves wrong - just moving it seems to be reporting mouse clicks to the app. Can I help somehow with this issue?

**Comment by BDisp** [comment]
Hi, yes, I know. Although stopping mouse repporting at exit, the terminal continues to report these movements and hangs the terminal. I think the problem is not from `.NET` because with `NetDriver` this does not happen. If you can help, I'll thank you very much, you do not even have to ask :-)

**Comment by Artemigos** [comment]
I'll try to take a look sometime this week, we'll see how that goes :)

**Comment by Artemigos** [comment]
"this week" became "next week" :)

I don't have much success so far, the mouse situation seems to be putting ncurses in a weird state - if I click with the left mouse button, then mouse movement starts reporting left mouse clicks, if I use the mouse wheel, then mouse movement starts reporting mouse wheel events, but much faster than before.

I tried to regenerate the `constants.cs` file for ncurses with the original code from mono-ncurses, the differences mostly weren't interesting except for maybe `Button4` codes (they are different) and `AllEvents` - it generated `0xfffffff` but there's `0x7ffffff` in code. I changed it, but I didn't notice any difference in functionality.

On the topic of interesting observations - the ncurses library gets loaded dynamically twice. I'm not sure if this causes any side effects yet, this might be the next thing I verify.

Long story short I'm worried that the native binding isn't completely correct, but I haven't found what exactly is wrong (yet).

**Comment by BDisp** [comment]
I finally fixed it, I hope :-) #931

**Comment by tig** [comment]
Leaving open until we verify.

**Comment by Artemigos** [comment]
For me, quitting the UICatalog now works a lot better, but still not completely correctly. After exiting, the terminal behaves as if only 25 lines of text could be visible on the screen, which is much less that is actually available. I can reproduce this in both alacritty and xterm. To maybe visualize better:

Normal terminal in the gui.cs repo runing `ll` three times:
![image](https://user-images.githubusercontent.com/4130682/116278583-ebe45c80-a786-11eb-9107-45541d2a6cf3.png)

The same situation right after quitting the UICatalog:
![image](https://user-images.githubusercontent.com/4130682/116278699-09b1c180-a787-11eb-8de6-390a12f89b0e.png)

The output just scrolls up as if it didn't have more lines available. Resizing the terminal fixes this immediately.

Also, it's starting to look like my problem with crazy mouse behavior is a different issue.

**Comment by BDisp** [comment]
Maybe is due to this:

`Console.Out.Write ("\x1b[1;25r");`

Can you please check it?

**Comment by Artemigos** [comment]
Commenting out this and the flush right after seems to fix it :D
The prompt after that is at the bottom at the terminal, but this is a nitpick:
![image](https://user-images.githubusercontent.com/4130682/116281364-efc5ae00-a789-11eb-9d7a-619d9bbd892b.png)

For comparison, this is what it looks like after I exit neovim:
![image](https://user-images.githubusercontent.com/4130682/116281614-2d2a3b80-a78a-11eb-8b42-84f114f14330.png)

Not sure if this is worth doing, though.

**Comment by BDisp** [comment]
Try commenting all this:

```cs
Console.Out.Write ("\x1b[2J");
Console.Out.Flush ();
Console.Out.Write ("\x1b[1;25r");
Console.Out.Flush ();
```

**Comment by Artemigos** [comment]
Perfect:
![image](https://user-images.githubusercontent.com/4130682/116282529-2ea83380-a78b-11eb-8a6f-7e07a2cce397.png)

**Comment by BDisp** [comment]
Ok thanks. I'll send a PR with that lines commented.

**Comment by tig** [comment]
This is great sluthing folks! 

This stuff is black-art level stuff. The more documentation on why these changes work, the better.

**Comment by BDisp** [comment]
I added this to the code:

```cs
// I'm commenting this because was used in a trying to fix the Linux hanging and forgot to exclude it.
// Clear and reset entire screen.
//Console.Out.Write ("\x1b[2J");
//Console.Out.Flush ();

// Reports current cursor line and column.
//Console.Out.Write ("\x1b[1;25r");
//Console.Out.Flush ();
```
